reference/matrix/fbcsr_kernels.cpp

```diff
@@ -0,0 +1,438 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/fbcsr_kernels.hpp"
+
+
+#include <algorithm>
+#include <iterator>
+#include <numeric>
+#include <utility>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/blockutils.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/base/allocator.hpp"
+#include "core/base/iterator_factory.hpp"
+#include "core/components/fixed_block.hpp"
+#include "core/components/prefix_sum.hpp"
+#include "core/matrix/fbcsr_builder.hpp"
+#include "reference/components/format_conversion.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The Compressed sparse row matrix format namespace.
+ * @ref Fbcsr
+ * @ingroup fbcsr
+ */
+namespace fbcsr {
+
+
+template <typename ValueType, typename IndexType>
+void spmv(const std::shared_ptr<const ReferenceExecutor> exec,
+          const matrix::Fbcsr<ValueType, IndexType> *const a,
+          const matrix::Dense<ValueType> *const b,
+          matrix::Dense<ValueType> *const c)
+{
+    const int bs = a->get_block_size();
+    const IndexType nvecs = static_cast<IndexType>(b->get_size()[1]);
+    const IndexType nbrows = a->get_num_block_rows();
+    auto row_ptrs = a->get_const_row_ptrs();
+    auto col_idxs = a->get_const_col_idxs();
+    auto vals = a->get_const_values();
+    const blockutils::DenseBlocksView<const ValueType, IndexType> avalues(
+        vals, bs, bs);
+
+    ValueType *const cvals = c->get_values();
+
+    for (IndexType ibrow = 0; ibrow < nbrows; ++ibrow) {
+        const IndexType crowblkend = (ibrow + 1) * bs * nvecs;
```

Comment:
In general I think using underscores makes the intent of the variables more clear. For example, `c_row_blk_end` instead of `crowblkend`